posthog/special_migrations/migrations/events_sample_by_migration.py

```diff
@@ -0,0 +1,103 @@
+from constance import config
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.events import EVENTS_TABLE, EVENTS_TABLE_MV_SQL, EVENTS_TABLE_SQL
+from posthog.constants import AnalyticsDBMS
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE
+from posthog.special_migrations.definition import SpecialMigrationDefinition, SpecialMigrationOperation
+from posthog.version_requirement import ServiceVersionRequirement
+
+ONE_DAY = 60 * 60 * 24
+
+TEMPORARY_TABLE_NAME = "temp_events"
+
+
+class Migration(SpecialMigrationDefinition):
+
+    description = "Events table migration for compatible sample by column."
+
+    posthog_min_version = "1.30.0"
+    posthog_max_version = "1.31.0"
+
+    service_version_requirements = [
+        ServiceVersionRequirement(service="clickhouse", supported_version=">=21.6.0,<21.7.0"),
+    ]
+
+    operations = [
+        SpecialMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"DETACH TABLE {EVENTS_TABLE_MV_SQL} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=f"ATTACH TABLE {EVENTS_TABLE_MV_SQL} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            side_effect=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", False),
+        ),
+        SpecialMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"CREATE TABLE IF NOT EXISTS {TEMPORARY_TABLE_NAME} AS {EVENTS_TABLE} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=f"DROP TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+        ),
+        SpecialMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+            INSERT INTO {TEMPORARY_TABLE_NAME}
+            SELECT * 
+            FROM {EVENTS_TABLE}""",
+            rollback=f"TRUNCATE TABLE {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+        ),
+        SpecialMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+                RENAME TABLE
+                    {CLICKHOUSE_DATABASE}.{EVENTS_TABLE} to {CLICKHOUSE_DATABASE}.{EVENTS_TABLE}_old,
+                    {CLICKHOUSE_DATABASE}.{TEMPORARY_TABLE_NAME} to {CLICKHOUSE_DATABASE}.{EVENTS_TABLE},
+                ON CLUSTER {CLICKHOUSE_CLUSTER}
+            """,
+            rollback=f"""
+                RENAME TABLE
+                    {CLICKHOUSE_DATABASE}.{EVENTS_TABLE} to {CLICKHOUSE_DATABASE}.{EVENTS_TABLE}_failed,
+                    {CLICKHOUSE_DATABASE}.{TEMPORARY_TABLE_NAME} to {CLICKHOUSE_DATABASE}.{EVENTS_TABLE},
```

Comment:
I had this bug in my example code. Fixed it earlier today. We want to restore the "old table" here

```suggestion
                    {CLICKHOUSE_DATABASE}.{EVENTS_TABLE}_old to {CLICKHOUSE_DATABASE}.{EVENTS_TABLE},
```